File: src/clients/node/src/wire_format_test.ts
/**
 * ArcherDB Node.js SDK - Wire Format Compatibility Tests
 *
 * These tests verify that the Node.js SDK produces wire-compatible
 * output with other language SDKs by testing against canonical test cases.
 */

import assert from 'assert'
import * as fs from 'fs'
import * as path from 'path'

import {
  // Constants
  LAT_MAX,
  LON_MAX,
  NANODEGREES_PER_DEGREE,
  MM_PER_METER,
  CENTIDEGREES_PER_DEGREE,
  BATCH_SIZE_MAX,
  QUERY_LIMIT_MAX,
  POLYGON_VERTICES_MAX,
  // Enums
  GeoEventFlags,
  GeoOperation,
  InsertGeoEventError,
  DeleteEntityError,
  // Conversion helpers
  degreesToNano,
  nanoToDegrees,
  metersToMm,
  mmToMeters,
  headingToCentidegrees,
  centidegreesToHeading,
  isValidLatitude,
  isValidLongitude,
  // Builder functions
  createGeoEvent,
  createRadiusQuery,
  createPolygonQuery,
} from './geo'

// Test data types
interface TestData {
  constants: {
    LAT_MAX: number
    LON_MAX: number
    NANODEGREES_PER_DEGREE: number
    MM_PER_METER: number
    CENTIDEGREES_PER_DEGREE: number
    BATCH_SIZE_MAX: number
    QUERY_LIMIT_MAX: number
    POLYGON_VERTICES_MAX: number
  }
  operation_codes: Record<string, number>
  geo_event_flags: Record<string, number>
  insert_result_codes: Record<string, number>
  delete_result_codes: Record<string, number>
  coordinate_conversions: Array<{
    description: string
    degrees: number
    expected_nanodegrees: number
  }>
  distance_conversions: Array<{
    description: string
    meters: number
    expected_mm: number
  }>
  heading_conversions: Array<{
    description: string
    degrees: number
    expected_centidegrees: number
  }>
  geo_events: Array<{
    description: string
    input: {
      entity_id: number
      latitude: number
      longitude: number
      correlation_id?: number
      user_data?: number
      group_id?: number
      altitude_m?: number
      velocity_mps?: number
      ttl_seconds?: number
      accuracy_m?: number
      heading?: number
      flags?: number
    }
    expected: {
      entity_id: number
      lat_nano: number
      lon_nano: number
      id: number
      timestamp: number
      correlation_id: number
      user_data: number
      group_id: number
      altitude_mm: number
      velocity_mms: number
      ttl_seconds: number
      accuracy_mm: number
      heading_cdeg: number
      flags: number
    }
  }>
  radius_queries: Array<{
    description: string
    input: {
      latitude: number
      longitude: number
      radius_m: number
      limit?: number
      timestamp_min?: number
      timestamp_max?: number
      group_id?: number
    }
    expected: {
      center_lat_nano: number
      center_lon_nano: number
      radius_mm: number
      limit: number
      timestamp_min: number
      timestamp_max: number
      group_id: number
    }
  }>
  polygon_queries: Array<{
    description: string
    input: {
      vertices: [number, number][]
      limit?: number
      timestamp_min?: number
      timestamp_max?: number
      group_id?: number
    }
    expected: {
      vertices: Array<{ lat_nano: number; lon_nano: number }>
      limit: number
      timestamp_min: number
      timestamp_max: number
      group_id: number
    }
  }>
  validation_cases: {
    invalid_latitudes: number[]
    invalid_longitudes: number[]
    valid_boundary_latitudes: number[]
    valid_boundary_longitudes: number[]
  }
}

// Load test data
function loadTestData(): TestData {
  const testDataPath = path.join(__dirname, '../../test-data/wire-format-test-cases.json')
  const data = fs.readFileSync(testDataPath, 'utf-8')
  return JSON.parse(data) as TestData
}

let testData: TestData
let passed = 0
let failed = 0

function test(name: string, fn: () => void): void {
  try {
    fn()
    console.log(`  ✓ ${name}`)
    passed++
  } catch (e) {
    console.log(`  ✗ ${name}`)
    console.log(`    Error: ${e instanceof Error ? e.message : e}`)
    failed++
  }
}

function assertEqual<T>(actual: T, expected: T, message: string): void {
  if (actual !== expected) {
    throw new Error(`${message}: expected ${expected}, got ${actual}`)
  }
}

function assertBigIntEqual(actual: bigint, expected: number | bigint, message: string): void {
  const expectedBigInt = typeof expected === 'bigint' ? expected : BigInt(expected)
  if (actual !== expectedBigInt) {
    throw new Error(`${message}: expected ${expectedBigInt}, got ${actual}`)
  }
}

// Test suites
function testConstants(): void {
  console.log('\nWire Format Constants:')

  test('LAT_MAX matches canonical value', () => {
    assertEqual(LAT_MAX, testData.constants.LAT_MAX, 'LAT_MAX')
  })

  test('LON_MAX matches canonical value', () => {
    assertEqual(LON_MAX, testData.constants.LON_MAX, 'LON_MAX')
  })

  test('NANODEGREES_PER_DEGREE matches canonical value', () => {
    assertBigIntEqual(NANODEGREES_PER_DEGREE, testData.constants.NANODEGREES_PER_DEGREE, 'NANODEGREES_PER_DEGREE')
  })

  test('MM_PER_METER matches canonical value', () => {
    assertEqual(MM_PER_METER, testData.constants.MM_PER_METER, 'MM_PER_METER')
  })

  test('CENTIDEGREES_PER_DEGREE matches canonical value', () => {
    assertEqual(CENTIDEGREES_PER_DEGREE, testData.constants.CENTIDEGREES_PER_DEGREE, 'CENTIDEGREES_PER_DEGREE')
  })

  test('BATCH_SIZE_MAX matches canonical value', () => {
    assertEqual(BATCH_SIZE_MAX, testData.constants.BATCH_SIZE_MAX, 'BATCH_SIZE_MAX')
  })

  test('QUERY_LIMIT_MAX matches canonical value', () => {
    assertEqual(QUERY_LIMIT_MAX, testData.constants.QUERY_LIMIT_MAX, 'QUERY_LIMIT_MAX')
  })

  test('POLYGON_VERTICES_MAX matches canonical value', () => {
    assertEqual(POLYGON_VERTICES_MAX, testData.constants.POLYGON_VERTICES_MAX, 'POLYGON_VERTICES_MAX')
  })
}

function testOperationCodes(): void {
  console.log('\nWire Format Operation Codes:')

  test('INSERT_EVENTS matches canonical value', () => {
    assertEqual(GeoOperation.insert_events, testData.operation_codes.INSERT_EVENTS, 'INSERT_EVENTS')
  })

  test('UPSERT_EVENTS matches canonical value', () => {
    assertEqual(GeoOperation.upsert_events, testData.operation_codes.UPSERT_EVENTS, 'UPSERT_EVENTS')
  })

  test('DELETE_ENTITIES matches canonical value', () => {
    assertEqual(GeoOperation.delete_entities, testData.operation_codes.DELETE_ENTITIES, 'DELETE_ENTITIES')
  })

  test('QUERY_UUID matches canonical value', () => {
    assertEqual(GeoOperation.query_uuid, testData.operation_codes.QUERY_UUID, 'QUERY_UUID')
  })

  test('QUERY_RADIUS matches canonical value', () => {
    assertEqual(GeoOperation.query_radius, testData.operation_codes.QUERY_RADIUS, 'QUERY_RADIUS')
  })

  test('QUERY_POLYGON matches canonical value', () => {
    assertEqual(GeoOperation.query_polygon, testData.operation_codes.QUERY_POLYGON, 'QUERY_POLYGON')
  })

  test('QUERY_LATEST matches canonical value', () => {
    assertEqual(GeoOperation.query_latest, testData.operation_codes.QUERY_LATEST, 'QUERY_LATEST')
  })
}

function testGeoEventFlags(): void {
  console.log('\nWire Format GeoEvent Flags:')

  test('NONE matches canonical value', () => {
    assertEqual(GeoEventFlags.none, testData.geo_event_flags.NONE, 'NONE')
  })

  test('LINKED matches canonical value', () => {
    assertEqual(GeoEventFlags.linked, testData.geo_event_flags.LINKED, 'LINKED')
  })

  test('IMPORTED matches canonical value', () => {
    assertEqual(GeoEventFlags.imported, testData.geo_event_flags.IMPORTED, 'IMPORTED')
  })

  test('STATIONARY matches canonical value', () => {
    assertEqual(GeoEventFlags.stationary, testData.geo_event_flags.STATIONARY, 'STATIONARY')
  })

  test('LOW_ACCURACY matches canonical value', () => {
    assertEqual(GeoEventFlags.low_accuracy, testData.geo_event_flags.LOW_ACCURACY, 'LOW_ACCURACY')
  })

  test('OFFLINE matches canonical value', () => {
    assertEqual(GeoEventFlags.offline, testData.geo_event_flags.OFFLINE, 'OFFLINE')
  })

  test('DELETED matches canonical value', () => {
    assertEqual(GeoEventFlags.deleted, testData.geo_event_flags.DELETED, 'DELETED')
  })
}

function testResultCodes(): void {
  console.log('\nWire Format Result Codes:')

  // Insert result codes
  test('Insert OK matches canonical value', () => {
    assertEqual(InsertGeoEventError.ok, testData.insert_result_codes.OK, 'OK')
  })

  test('Insert LINKED_EVENT_FAILED matches canonical value', () => {
    assertEqual(InsertGeoEventError.linked_event_failed, testData.insert_result_codes.LINKED_EVENT_FAILED, 'LINKED_EVENT_FAILED')
  })

  test('Insert INVALID_COORDINATES matches canonical value', () => {
    assertEqual(InsertGeoEventError.invalid_coordinates, testData.insert_result_codes.INVALID_COORDINATES, 'INVALID_COORDINATES')
  })

  test('Insert EXISTS matches canonical value', () => {
    assertEqual(InsertGeoEventError.exists, testData.insert_result_codes.EXISTS, 'EXISTS')
  })

  // Delete result codes
  test('Delete OK matches canonical value', () => {
    assertEqual(DeleteEntityError.ok, testData.delete_result_codes.OK, 'OK')
  })

  test('Delete ENTITY_NOT_FOUND matches canonical value', () => {
    assertEqual(DeleteEntityError.entity_not_found, testData.delete_result_codes.ENTITY_NOT_FOUND, 'ENTITY_NOT_FOUND')
  })
}

function testCoordinateConversions(): void {
  console.log('\nWire Format Coordinate Conversions:')

  for (const testCase of testData.coordinate_conversions) {
    test(`degreesToNano: ${testCase.description}`, () => {
      const result = degreesToNano(testCase.degrees)
      assertBigIntEqual(result, testCase.expected_nanodegrees, testCase.description)
    })
  }

  // Roundtrip test
  test('nanoToDegrees roundtrip maintains precision', () => {
    for (const testCase of testData.coordinate_conversions) {
      const nano = BigInt(testCase.expected_nanodegrees)
      const degrees = nanoToDegrees(nano)
      const backToNano = degreesToNano(degrees)
      assertBigIntEqual(backToNano, testCase.expected_nanodegrees, `Roundtrip: ${testCase.description}`)
    }
  })
}

function testDistanceConversions(): void {
  console.log('\nWire Format Distance Conversions:')

  for (const testCase of testData.distance_conversions) {
    test(`metersToMm: ${testCase.description}`, () => {
      const result = metersToMm(testCase.meters)
      assertEqual(result, testCase.expected_mm, testCase.description)
    })
  }
}

function testHeadingConversions(): void {
  console.log('\nWire Format Heading Conversions:')

  for (const testCase of testData.heading_conversions) {
    test(`headingToCentidegrees: ${testCase.description}`, () => {
      const result = headingToCentidegrees(testCase.degrees)
      assertEqual(result, testCase.expected_centidegrees, testCase.description)
    })
  }
}

function testGeoEvents(): void {
  console.log('\nWire Format GeoEvent Creation:')

  for (const testCase of testData.geo_events) {
    test(`createGeoEvent: ${testCase.description}`, () => {
      const input = testCase.input
      const expected = testCase.expected

      const event = createGeoEvent({
        entity_id: BigInt(input.entity_id),
        latitude: input.latitude,
        longitude: input.longitude,
        correlation_id: input.correlation_id !== undefined ? BigInt(input.correlation_id) : undefined,
        user_data: input.user_data !== undefined ? BigInt(input.user_data) : undefined,
        group_id: input.group_id !== undefined ? BigInt(input.group_id) : undefined,
        altitude_m: input.altitude_m,
        velocity_mps: input.velocity_mps,
        ttl_seconds: input.ttl_seconds,
        accuracy_m: input.accuracy_m,
        heading: input.heading,
        flags: input.flags,
      })

      assertBigIntEqual(event.entity_id, expected.entity_id, 'entity_id')
      assertBigIntEqual(event.lat_nano, expected.lat_nano, 'lat_nano')
      assertBigIntEqual(event.lon_nano, expected.lon_nano, 'lon_nano')
      assertBigIntEqual(event.id, expected.id, 'id')
      assertBigIntEqual(event.timestamp, expected.timestamp, 'timestamp')
      assertBigIntEqual(event.correlation_id, expected.correlation_id, 'correlation_id')
      assertBigIntEqual(event.user_data, expected.user_data, 'user_data')
      assertBigIntEqual(event.group_id, expected.group_id, 'group_id')
      assertEqual(event.altitude_mm, expected.altitude_mm, 'altitude_mm')
      assertEqual(event.velocity_mms, expected.velocity_mms, 'velocity_mms')
      assertEqual(event.ttl_seconds, expected.ttl_seconds, 'ttl_seconds')
      assertEqual(event.accuracy_mm, expected.accuracy_mm, 'accuracy_mm')
      assertEqual(event.heading_cdeg, expected.heading_cdeg, 'heading_cdeg')
      assertEqual(event.flags, expected.flags, 'flags')
    })
  }
}

function testRadiusQueries(): void {
  console.log('\nWire Format Radius Query Creation:')

  for (const testCase of testData.radius_queries) {
    test(`createRadiusQuery: ${testCase.description}`, () => {
      const input = testCase.input
      const expected = testCase.expected

      const query = createRadiusQuery({
        latitude: input.latitude,
        longitude: input.longitude,
        radius_m: input.radius_m,
        limit: input.limit,
        timestamp_min: input.timestamp_min !== undefined ? BigInt(input.timestamp_min) : undefined,
        timestamp_max: input.timestamp_max !== undefined ? BigInt(input.timestamp_max) : undefined,
        group_id: input.group_id !== undefined ? BigInt(input.group_id) : undefined,
      })

      assertBigIntEqual(query.center_lat_nano, expected.center_lat_nano, 'center_lat_nano')
      assertBigIntEqual(query.center_lon_nano, expected.center_lon_nano, 'center_lon_nano')
      assertEqual(query.radius_mm, expected.radius_mm, 'radius_mm')
      assertEqual(query.limit, expected.limit, 'limit')
      assertBigIntEqual(query.timestamp_min, expected.timestamp_min, 'timestamp_min')
      assertBigIntEqual(query.timestamp_max, expected.timestamp_max, 'timestamp_max')
      assertBigIntEqual(query.group_id, expected.group_id, 'group_id')
    })
  }
}

function testPolygonQueries(): void {
  console.log('\nWire Format Polygon Query Creation:')

  for (const testCase of testData.polygon_queries) {
    test(`createPolygonQuery: ${testCase.description}`, () => {
      const input = testCase.input
      const expected = testCase.expected

      // Convert JSON [lat, lon] arrays to tuples
      const vertices: Array<[number, number]> = input.vertices.map(
        (v: number[]) => [v[0], v[1]] as [number, number]
      )

      const query = createPolygonQuery({
        vertices,
        limit: input.limit,
        timestamp_min: input.timestamp_min !== undefined ? BigInt(input.timestamp_min) : undefined,
        timestamp_max: input.timestamp_max !== undefined ? BigInt(input.timestamp_max) : undefined,
        group_id: input.group_id !== undefined ? BigInt(input.group_id) : undefined,
      })

      assertEqual(query.vertices.length, expected.vertices.length, 'vertex count')
      for (let i = 0; i < expected.vertices.length; i++) {
        assertBigIntEqual(query.vertices[i].lat_nano, expected.vertices[i].lat_nano, `vertex ${i} lat_nano`)
        assertBigIntEqual(query.vertices[i].lon_nano, expected.vertices[i].lon_nano, `vertex ${i} lon_nano`)
      }
      assertEqual(query.limit, expected.limit, 'limit')
      assertBigIntEqual(query.timestamp_min, expected.timestamp_min, 'timestamp_min')
      assertBigIntEqual(query.timestamp_max, expected.timestamp_max, 'timestamp_max')
      assertBigIntEqual(query.group_id, expected.group_id, 'group_id')
    })
  }
}

function testValidation(): void {
  console.log('\nWire Format Validation:')

  test('Invalid latitudes are rejected', () => {
    for (const lat of testData.validation_cases.invalid_latitudes) {
      assert(!isValidLatitude(lat), `Latitude ${lat} should be invalid`)
    }
  })

  test('Invalid longitudes are rejected', () => {
    for (const lon of testData.validation_cases.invalid_longitudes) {
      assert(!isValidLongitude(lon), `Longitude ${lon} should be invalid`)
    }
  })

  test('Valid boundary latitudes are accepted', () => {
    for (const lat of testData.validation_cases.valid_boundary_latitudes) {
      assert(isValidLatitude(lat), `Latitude ${lat} should be valid`)
    }
  })

  test('Valid boundary longitudes are accepted', () => {
    for (const lon of testData.validation_cases.valid_boundary_longitudes) {
      assert(isValidLongitude(lon), `Longitude ${lon} should be valid`)
    }
  })
}

// Main
function main(): void {
  console.log('ArcherDB Node.js SDK - Wire Format Compatibility Tests')
  console.log('='.repeat(60))

  try {
    testData = loadTestData()
  } catch (e) {
    console.error('Failed to load test data:', e)
    process.exit(1)
  }

  testConstants()
  testOperationCodes()
  testGeoEventFlags()
  testResultCodes()
  testCoordinateConversions()
  testDistanceConversions()
  testHeadingConversions()
  testGeoEvents()
  testRadiusQueries()
  testPolygonQueries()
  testValidation()

  console.log('\n' + '='.repeat(60))
  console.log(`WIRE FORMAT COMPATIBILITY: ${passed} passed, ${failed} failed`)

  if (failed > 0) {
    process.exit(1)
  }
}

main()

File: src/clients/java/src/test/java/com/archerdb/geo/WireFormatTest.java
package com.archerdb.geo;

import static org.junit.jupiter.api.Assertions.*;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.stream.Stream;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;
import org.junit.jupiter.params.provider.Arguments;

/**
 * Wire format compatibility tests.
 *
 * Loads canonical test data from wire-format-test-cases.json and verifies the Java SDK produces
 * compatible output with other language SDKs.
 */
class WireFormatTest {

    private static JsonObject testData;

    @BeforeAll
    static void loadTestData() throws IOException {
        Path testDataPath = findTestDataPath();
        String json = Files.readString(testDataPath);
        testData = new Gson().fromJson(json, JsonObject.class);
    }

    private static Path findTestDataPath() {
        // Test runs from src/clients/java/, so ../test-data is the correct relative path
        Path[] candidates = {Paths.get("../test-data/wire-format-test-cases.json"),
                Paths.get("../../test-data/wire-format-test-cases.json"),
                Paths.get("src/clients/test-data/wire-format-test-cases.json"),
                Paths.get("test-data/wire-format-test-cases.json"),};

        for (Path path : candidates) {
            if (Files.exists(path)) {
                return path;
            }
        }

        String projectRoot = System.getProperty("user.dir");
        for (Path path : candidates) {
            Path fullPath = Paths.get(projectRoot).resolve(path);
            if (Files.exists(fullPath)) {
                return fullPath;
            }
        }

        throw new RuntimeException("Could not find wire-format-test-cases.json");
    }

    // ========================================================================
    // Constants Tests
    // ========================================================================

    @Test
    @DisplayName("Wire format constants match canonical values")
    void testConstants() {
        JsonObject constants = testData.getAsJsonObject("constants");

        assertEquals(constants.get("LAT_MAX").getAsDouble(), CoordinateUtils.LAT_MAX, "LAT_MAX");
        assertEquals(constants.get("LON_MAX").getAsDouble(), CoordinateUtils.LON_MAX, "LON_MAX");
        assertEquals(constants.get("NANODEGREES_PER_DEGREE").getAsLong(),
                CoordinateUtils.NANODEGREES_PER_DEGREE, "NANODEGREES_PER_DEGREE");
        assertEquals(constants.get("MM_PER_METER").getAsInt(), CoordinateUtils.MM_PER_METER,
                "MM_PER_METER");
        assertEquals(constants.get("CENTIDEGREES_PER_DEGREE").getAsInt(),
                CoordinateUtils.CENTIDEGREES_PER_DEGREE, "CENTIDEGREES_PER_DEGREE");
        assertEquals(constants.get("BATCH_SIZE_MAX").getAsInt(), CoordinateUtils.BATCH_SIZE_MAX,
                "BATCH_SIZE_MAX");
        assertEquals(constants.get("QUERY_LIMIT_MAX").getAsInt(), CoordinateUtils.QUERY_LIMIT_MAX,
                "QUERY_LIMIT_MAX");
        assertEquals(constants.get("POLYGON_VERTICES_MAX").getAsInt(),
                CoordinateUtils.POLYGON_VERTICES_MAX, "POLYGON_VERTICES_MAX");
    }

    // ========================================================================
    // GeoEvent Flags Tests
    // ========================================================================

    @Test
    @DisplayName("Wire format GeoEvent flags match canonical values")
    void testGeoEventFlags() {
        JsonObject flags = testData.getAsJsonObject("geo_event_flags");

        assertEquals(flags.get("NONE").getAsInt(), GeoEventFlags.NONE.getValue(), "NONE");
        assertEquals(flags.get("LINKED").getAsInt(), GeoEventFlags.LINKED.getValue(), "LINKED");
        assertEquals(flags.get("IMPORTED").getAsInt(), GeoEventFlags.IMPORTED.getValue(),
                "IMPORTED");
        assertEquals(flags.get("STATIONARY").getAsInt(), GeoEventFlags.STATIONARY.getValue(),
                "STATIONARY");
        assertEquals(flags.get("LOW_ACCURACY").getAsInt(), GeoEventFlags.LOW_ACCURACY.getValue(),
                "LOW_ACCURACY");
        assertEquals(flags.get("OFFLINE").getAsInt(), GeoEventFlags.OFFLINE.getValue(), "OFFLINE");
        assertEquals(flags.get("DELETED").getAsInt(), GeoEventFlags.DELETED.getValue(), "DELETED");
    }

    // ========================================================================
    // Result Codes Tests
    // ========================================================================

    @Test
    @DisplayName("Wire format insert result codes match canonical values")
    void testInsertResultCodes() {
        JsonObject codes = testData.getAsJsonObject("insert_result_codes");

        assertEquals(codes.get("OK").getAsInt(), InsertGeoEventResult.OK.getCode(), "OK");
        assertEquals(codes.get("INVALID_COORDINATES").getAsInt(),
                InsertGeoEventResult.INVALID_COORDINATES.getCode(), "INVALID_COORDINATES");
        assertEquals(codes.get("EXISTS").getAsInt(), InsertGeoEventResult.EXISTS.getCode(),
                "EXISTS");
        assertEquals(codes.get("LINKED_EVENT_FAILED").getAsInt(),
                InsertGeoEventResult.LINKED_EVENT_FAILED.getCode(), "LINKED_EVENT_FAILED");
    }

    // ========================================================================
    // Coordinate Conversion Tests
    // ========================================================================

    static Stream<Arguments> coordinateConversionsProvider() {
        JsonArray conversions = testData.getAsJsonArray("coordinate_conversions");
        return conversions.asList().stream().map(e -> {
            JsonObject obj = e.getAsJsonObject();
            return Arguments.of(obj.get("description").getAsString(),
                    obj.get("degrees").getAsDouble(), obj.get("expected_nanodegrees").getAsLong());
        });
    }

    @ParameterizedTest(name = "{0}")
    @MethodSource("coordinateConversionsProvider")
    @DisplayName("Wire format coordinate conversions")
    void testCoordinateConversions(String description, double degrees, long expectedNano) {
        long result = CoordinateUtils.degreesToNano(degrees);
        assertEquals(expectedNano, result, description);
    }

    @ParameterizedTest(name = "Roundtrip: {0}")
    @MethodSource("coordinateConversionsProvider")
    @DisplayName("Wire format coordinate roundtrip")
    void testCoordinateRoundtrip(String description, double degrees, long expectedNano) {
        double back = CoordinateUtils.nanoToDegrees(expectedNano);
        long roundTrip = CoordinateUtils.degreesToNano(back);
        assertEquals(expectedNano, roundTrip, "Roundtrip: " + description);
    }

    // ========================================================================
    // Distance Conversion Tests
    // ========================================================================

    static Stream<Arguments> distanceConversionsProvider() {
        JsonArray conversions = testData.getAsJsonArray("distance_conversions");
        return conversions.asList().stream().map(e -> {
            JsonObject obj = e.getAsJsonObject();
            return Arguments.of(obj.get("description").getAsString(),
                    obj.get("meters").getAsDouble(), obj.get("expected_mm").getAsInt());
        });
    }

    @ParameterizedTest(name = "{0}")
    @MethodSource("distanceConversionsProvider")
    @DisplayName("Wire format distance conversions")
    void testDistanceConversions(String description, double meters, int expectedMm) {
        int result = CoordinateUtils.metersToMm(meters);
        assertEquals(expectedMm, result, description);
    }

    // ========================================================================
    // Heading Conversion Tests
    // ========================================================================

    static Stream<Arguments> headingConversionsProvider() {
        JsonArray conversions = testData.getAsJsonArray("heading_conversions");
        return conversions.asList().stream().map(e -> {
            JsonObject obj = e.getAsJsonObject();
            return Arguments.of(obj.get("description").getAsString(),
                    obj.get("degrees").getAsDouble(), obj.get("expected_centidegrees").getAsInt());
        });
    }

    @ParameterizedTest(name = "{0}")
    @MethodSource("headingConversionsProvider")
    @DisplayName("Wire format heading conversions")
    void testHeadingConversions(String description, double degrees, int expectedCdeg) {
        short result = CoordinateUtils.headingToCentidegrees(degrees);
        assertEquals((short) expectedCdeg, result, description);
    }

    // ========================================================================
    // GeoEvent Creation Tests
    // ========================================================================

    static Stream<Arguments> geoEventsProvider() {
        JsonArray events = testData.getAsJsonArray("geo_events");
        return events.asList().stream().map(e -> {
            JsonObject obj = e.getAsJsonObject();
            return Arguments.of(obj.get("description").getAsString(), obj.getAsJsonObject("input"),
                    obj.getAsJsonObject("expected"));
        });
    }

    @ParameterizedTest(name = "{0}")
    @MethodSource("geoEventsProvider")
    @DisplayName("Wire format GeoEvent creation")
    void testGeoEventCreation(String description, JsonObject input, JsonObject expected) {
        GeoEvent.Builder builder = new GeoEvent.Builder()
                .setEntityId(UInt128.fromLong(input.get("entity_id").getAsLong()))
                .setLatitude(input.get("latitude").getAsDouble())
                .setLongitude(input.get("longitude").getAsDouble());

        if (input.has("correlation_id") && input.get("correlation_id").getAsLong() != 0) {
            builder.setCorrelationId(UInt128.fromLong(input.get("correlation_id").getAsLong()));
        }
        if (input.has("user_data") && input.get("user_data").getAsLong() != 0) {
            builder.setUserData(UInt128.fromLong(input.get("user_data").getAsLong()));
        }
        if (input.has("group_id") && input.get("group_id").getAsLong() != 0) {
            builder.setGroupId(input.get("group_id").getAsLong());
        }
        if (input.has("altitude_m") && input.get("altitude_m").getAsDouble() != 0) {
            builder.setAltitude(input.get("altitude_m").getAsDouble());
        }
        if (input.has("velocity_mps") && input.get("velocity_mps").getAsDouble() != 0) {
            builder.setVelocity(input.get("velocity_mps").getAsDouble());
        }
        if (input.has("ttl_seconds") && input.get("ttl_seconds").getAsInt() != 0) {
            builder.setTtlSeconds(input.get("ttl_seconds").getAsInt());
        }
        if (input.has("accuracy_m") && input.get("accuracy_m").getAsDouble() != 0) {
            builder.setAccuracy(input.get("accuracy_m").getAsDouble());
        }
        if (input.has("heading") && input.get("heading").getAsDouble() != 0) {
            builder.setHeading(input.get("heading").getAsDouble());
        }
        if (input.has("flags") && input.get("flags").getAsInt() != 0) {
            builder.setFlags((short) input.get("flags").getAsInt());
        }

        GeoEvent event = builder.build();

        // Verify fields
        assertEquals(expected.get("entity_id").getAsLong(), event.getEntityId().getLo(),
                "entity_id");
        assertEquals(expected.get("lat_nano").getAsLong(), event.getLatNano(), "lat_nano");
        assertEquals(expected.get("lon_nano").getAsLong(), event.getLonNano(), "lon_nano");
        // ID should be 0 before prepare
        assertEquals(0L, event.getId().getLo(), "id should be 0 before prepare");
        assertEquals(0L, event.getId().getHi(), "id high should be 0 before prepare");
        // Timestamp should be 0 (server-assigned)
        assertEquals(expected.get("timestamp").getAsLong(), event.getTimestamp(), "timestamp");
        assertEquals(expected.get("correlation_id").getAsLong(), event.getCorrelationId().getLo(),
                "correlation_id");
        assertEquals(expected.get("user_data").getAsLong(), event.getUserData().getLo(),
                "user_data");
        assertEquals(expected.get("group_id").getAsLong(), event.getGroupId(), "group_id");
        assertEquals(expected.get("altitude_mm").getAsInt(), event.getAltitudeMm(), "altitude_mm");
        assertEquals(expected.get("velocity_mms").getAsInt(), event.getVelocityMms(),
                "velocity_mms");
        assertEquals(expected.get("ttl_seconds").getAsInt(), event.getTtlSeconds(), "ttl_seconds");
        assertEquals(expected.get("accuracy_mm").getAsInt(), event.getAccuracyMm(), "accuracy_mm");
        assertEquals(expected.get("heading_cdeg").getAsInt(), (int) event.getHeadingCdeg(),
                "heading_cdeg");
        assertEquals(expected.get("flags").getAsInt(), (int) event.getFlags(), "flags");
    }

    // ========================================================================
    // Validation Tests
    // ========================================================================

    @Test
    @DisplayName("Invalid latitudes are rejected")
    void testInvalidLatitudes() {
        JsonArray invalids =
                testData.getAsJsonObject("validation_cases").getAsJsonArray("invalid_latitudes");
        for (var elem : invalids) {
            double lat = elem.getAsDouble();
            assertFalse(CoordinateUtils.isValidLatitude(lat),
                    "Latitude " + lat + " should be invalid");
        }
    }

    @Test
    @DisplayName("Invalid longitudes are rejected")
    void testInvalidLongitudes() {
        JsonArray invalids =
                testData.getAsJsonObject("validation_cases").getAsJsonArray("invalid_longitudes");
        for (var elem : invalids) {
            double lon = elem.getAsDouble();
            assertFalse(CoordinateUtils.isValidLongitude(lon),
                    "Longitude " + lon + " should be invalid");
        }
    }

    @Test
    @DisplayName("Valid boundary latitudes are accepted")
    void testValidBoundaryLatitudes() {
        JsonArray valids = testData.getAsJsonObject("validation_cases")
                .getAsJsonArray("valid_boundary_latitudes");
        for (var elem : valids) {
            double lat = elem.getAsDouble();
            assertTrue(CoordinateUtils.isValidLatitude(lat),
                    "Latitude " + lat + " should be valid");
        }
    }

    @Test
    @DisplayName("Valid boundary longitudes are accepted")
    void testValidBoundaryLongitudes() {
        JsonArray valids = testData.getAsJsonObject("validation_cases")
                .getAsJsonArray("valid_boundary_longitudes");
        for (var elem : valids) {
            double lon = elem.getAsDouble();
            assertTrue(CoordinateUtils.isValidLongitude(lon),
                    "Longitude " + lon + " should be valid");
        }
    }
}
